File: staircase.ts
/**
 * Prints a staircase of size n. The staircase is right-aligned.
 * @param n - the size of the staircase
 */
function staircase(n: number): void {
  // loop trough each step of the staircase
  for (let step = 1; step <= n; step++) {
    // each line has n - step spaces and step # symbols
    // last line will have n - n (0) spaces and n # symbols
    // NOTE: console.log automatically adds a newline character - check the golang version for a version without it
    console.log(" ".repeat(n - step) + "#".repeat(step));
  }
}

const nStr = prompt("Enter the size for the staircase: ");
const n = parseInt(nStr || "not a number", 10);

if (isNaN(n)) {
  console.error("No number provided.");
  Deno.exit(1);
}

staircase(n);

File: go/staircase.go
package main

import (
	"fmt"
	"strings"
)

// Each step has a number of '#' characters equal to the step number (i)
// As such, each step has n-i spaces followed by i '#' characters.
func Staircase(n int) {
	// loop through each step
	for i := 1; i <= n; i++ {
		// create a string with n-i spaces followed by i '#' characters
		var s string = strings.Repeat(" ", n-i) + strings.Repeat("#", i)

		// add a newline character if this is not the last step
		if i != n {
			s += "\n"
		}

		fmt.Print(s)
	}
}

func main() {
	var n int
	fmt.Print("Enter the number of rows for the staircase: ")

	_, err := fmt.Scan(&n)
	if err != nil {
		fmt.Println("Error reading input")
		return
	}

	fmt.Println("The staircase is:")
	Staircase(n)
}
